[HDR] Implement reliable way to detect an image hasHDRContent() or not

shallawa opened the pull request:
#### f798905acff7679506cc2693f0c572ee4fcfc13c
<pre>
[HDR] Implement reliable way to detect an image hasHDRContent() or not
<a href="https://bugs.webkit.org/show_bug.cgi?id=291431">https://bugs.webkit.org/show_bug.cgi?id=291431</a>
<a href="https://rdar.apple.com/149071262">rdar://149071262</a>

Reviewed by Simon Fraser.

It turned out using the CGImageSource frame &quot;Headroom&quot; metadata property is not
a reliable way to detect whether an image has HDR contents or not.

An image is considered to be HDR if one of these conditions is true:

1. The colorSpace of the primary frame usesITUR_2100TF()
2. The ImageDecoder detects the image has a gain map

To have the async image decoding workflow works as expected, destroy the primary
frame after getting its colorSpace only if it has not been used yet.

* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
(WebCore::BitmapImageDescriptor::hasHDRGainMap const):
(WebCore::BitmapImageDescriptor::hasHDRColorSpace const):
(WebCore::BitmapImageDescriptor::headroom const): Deleted.
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::usesITUR_2100TF const):
* Source/WebCore/platform/graphics/DestinationColorSpace.h:
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::hasHDRGainMap const):
(WebCore::ImageSource::shouldDecodeToHDR const):
* Source/WebCore/platform/graphics/NativeImageSource.h:

Canonical link: <a href="https://commits.webkit.org/294390@main">https://commits.webkit.org/294390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8fc43aa30feddb63ec57a2ba7eefdc1f0c903f8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101659 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11642 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106817 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52293 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103699 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21635 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29827 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77418 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34447 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104666 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16716 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91807 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57755 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16542 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9824 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/51648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/86409 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/9901 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109171 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28792 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21203 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29153 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88008 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/85963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21871 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30711 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8425 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22955 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28720 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34009 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28531 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31854 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30090 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
```diff
@@ -163,14 +163,22 @@ std::optional<Color> BitmapImageDescriptor::singlePixelSolidColor() const
+        return colorSpace().usesITUR_2100TF();
+
+    bool hasHDRContent = colorSpace().usesITUR_2100TF();
+    m_source->destroyNativeImageAtIndex(m_source->primaryFrameIndex());
```
There's an implicit assumption here that reading `colorSpace()` allocated the frame (because that's what `NativeImageSource` does. But that seems like an implementation detail of `NativeImageSource `, so this frame destruction should move there.

shallawa replied:
Destroying the frame here is unnecessary. But we need to fix the async image decoding code path such that it ignores this frame and treat it as if it did not exist. 

It seems ImageIO caches the frame also. So if we call `CGImageSourceCreateImageAtIndex()` ImageIO will return the cached frame without having to create a new one.

I added a FIXME comment and I will address fixing the async image decoding issue in a separate PR.

shallawa committed:
[HDR] Implement reliable way to detect an image hasHDRContent() or not
https://bugs.webkit.org/show_bug.cgi?id=291431
rdar://149071262

Reviewed by Simon Fraser.

It turned out using the CGImageSource frame "Headroom" metadata property is not
a reliable way to detect whether an image has HDR contents or not.

An image is considered to be HDR if one of these conditions is true:

1. The colorSpace of the primary frame usesITUR_2100TF()
2. The ImageDecoder detects the image has a gain map

To have the async image decoding workflow works as expected, destroy the primary
frame after getting its colorSpace only if it has not been used yet.

* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
(WebCore::BitmapImageDescriptor::hasHDRGainMap const):
(WebCore::BitmapImageDescriptor::hasHDRColorSpace const):
(WebCore::BitmapImageDescriptor::headroom const): Deleted.
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::usesITUR_2100TF const):
* Source/WebCore/platform/graphics/DestinationColorSpace.h:
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::hasHDRGainMap const):
(WebCore::ImageSource::shouldDecodeToHDR const):
* Source/WebCore/platform/graphics/NativeImageSource.h:

Canonical link: https://commits.webkit.org/294390@main
